vim: Implement Go To Previous Word End

vbhavsar opened the pull request:
Activated by keystrokes g-e.



Release Notes:

- vim: Added `ge` and `gE` for go to Previous Word End.

vbhavsar committed:
vim: Implement Previous Word End

Activated by keystrokes g-e.

vbhavsar commented on crates/vim/src/motion.rs:
```diff
@@ -1115,6 +1128,29 @@ fn window_bottom(
     }
 }
 
+fn previous_word_end(
```
@ConradIrwin This is how I've implemented this feature right now but the `is_boundary` function is proving to be a bit limiting for certain cases (see failing test case). I think to deal with corner cases I might need to write completely custom logic and not use `find_preceding_boundary`. Thoughts?

ConradIrwin replied:
I am totally fine if a custom loop is easier to write! The main thing that `find_preceding_boundary` gives you is the utf8 offset calculation for multibyte chars, but that's easy to copy.

What is the edge-case you're struggling with?

(as always, happy to help if you want to book time: https://calendly.com/conradirwin/pairing)

ConradIrwin replied:
Looking more at the code, would it be better to do more in the `find_preceding_boundary` so we don't need to do the previous_word_start/next_word_end after it's done. It looks like we have the same bug in `previous_word_start` too, so fixing that might help.

From the vim help:

```
A word consists of a sequence of letters, digits and underscores, or a
sequence of other non-blank characters, separated with white space (spaces,
tabs, <EOL>).  This can be changed with the 'iskeyword' option.  An empty line
is also considered to be a word.
```

So it seems like at a first pass a word end is the preceding boundary where `left is alphanumeric and right is not, OR left is punctuation and right is alphanumeric, OR left and right are both newlines`?

vbhavsar committed:
Remove println

ConradIrwin commented:
@vbhavsar are you still interested in pairing on this?

vbhavsar committed:
Fix movement

vbhavsar commented:
Hey @ConradIrwin. Sorry this was sitting for a bit. I changed the code a bit and it works for the basic use case of words separated by spaces. It doesn't quite work when two character kinds are next to each other (like `:ABC`).

It feels like a lot more work needs to be done to support the more advanced use cases.

vbhavsar committed:
Clean up

ConradIrwin committed:
Tidy up `g e` implementation

Go back to using find_boundary, but start one further to the right so
that we can find the immediately preceding character.

ConradIrwin committed:
Add gE too (and fix some clipping bugs)

ConradIrwin commented:
@vbhavsar thanks for getting this started! I took a bit of a more detailed look, and I see why you were struggling with the `found_boundaries` code.

I've gone ahead and fixed that, and implemented the remaining edge cases that I could find!

vbhavsar commented:
@ConradIrwin That's great. Thanks!
